src/persist/src/nemesis/generator.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use rand::prelude::SmallRng;
+use rand::{Rng, SeedableRng};
+
+use crate::nemesis::{
+    Input, ReadSnapshotReq, Req, ReqId, SealReq, SnapshotId, TakeSnapshotReq, WriteReq,
+};
+
+pub struct Generator {
+    seed: u64,
+    req_id: u64,
+    snap_id: u64,
+}
+
+// TODO: Make Generator configurable, so individual tests can be written to
+// stress interesting combinations of random traffic.
+impl Generator {
+    pub fn from_seed(seed: u64) -> Self {
+        Generator {
+            seed,
+            req_id: 0,
+            snap_id: 0,
+        }
+    }
+
+    pub fn gen(&mut self) -> Input {
+        let req_id = ReqId(self.req_id);
+        self.req_id += 1;
+        let mut rng = SmallRng::seed_from_u64(self.seed + req_id.0);
+        let req = loop {
+            break match rng.gen_range(0..=3) {
+                0 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let key = rng.gen_range('a'..'e').to_string();
+                    let ts = rng.gen_range(0..100);
```

Comment:
This will distribute timestamps uniformly over the space, but I thought timestamps are monotonically increasing? Can you get an event for a previous timestamp after a more recent one has been seen?

Reply:
>  A user with some sort of dataflow input handle gets a "capability" to introduce data at any timestamp they like. As they go along, they "downgrade" the capability to a higher timestamp, which is a promise that they're done introducing data below that timestamp.

Thanks for the explanation, so before you retire timestamp X, you are allowed to introduce events for X , X - 1 and all the way down to timestamp 0. Good to know, I thought timely will straight up panic on you if you introduced an event for X - 1 or lower. Will keep that in mind when it comes to testing future stuff.

Reply:
Right. Before you've retired anything, you can introduce data at [0, inf). Then if you retire X, you're allowed to introduce data at [X, inf). I'm actually not sure what timely does here if you use it wrong, it very well might panic. However, in persist, we already have to handle errors resulting from io failures, etc, so we may as well more gracefully handle incorrect API usage and return an error.

Reply:
timely panics if you try to downgrade capabilities incorrectly